components/executor/executor_impl.go

```diff
@@ -86,48 +96,81 @@ func (e *executorImpl) Delete(jobId string) error {
 	return ErrJobNotFound
 }
 
-// JobCh returns the channel used to receive jobs.
-func (e *executorImpl) JobCh() chan *jobmodels.Job {
-	return e.jobCh
-}
-
 // Close closes the executor and waits for all the jobs to finish executing.
 func (e *executorImpl) Close() {
 	e.rw.Lock()
 	defer e.rw.Unlock()
 
 	e.isClosed = true
 
-	// Wait for all the jobs to execute
-	for len(e.jobs) > 0 {
-		time.Sleep(1 * time.Second)
-	}
+	// Stop dispatcher and wait for it to stop
+	e.stopDispatcher()
+	e.wgDispacther.Wait()
 
+	// close dispatcher channel
+	// so that executor go-routines can terminate
 	close(e.jobCh)
 }
 
-func (e *executorImpl) scheduleJob(jobId string, version int, triggerTime time.Time) {
-	time.AfterFunc(time.Until(triggerTime), func() {
+func (e *executorImpl) nextTick() {
+	var now = time.Now()
+	for {
+		jentry := e.jobQueue.nextJob()
```

Comment:
Since we are having our own heap implementation, we can have a method which returns the time for the job with the least time. This way, we dont need to pop and insert back again. 

We couldnt do this in redis sorted set. But we can implement our own method.

Reply:
As per the heap interface provided by Golang, We can't iterate the heap's underlying array. Push and Pop are two fundamental methods provided by Heap ADT in general. 

Also, our Push and Pop implementation is quite fast.